### PR TITLE
[SPARK-28322][SQL] Add support to Decimal type for integral divide

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

The expression `IntegralDivide`, which corresponds to the `div` operator, support only integral type. Postgres, though, allows it to work also with decimals.

The PR adds the support to decimal operands for this operation in order to have feature parity with postgres.

## How was this patch tested?

added UTs

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by mgaido91** [comment]
cc @cloud-fan @gatorsmile

**Comment by mgaido91** [comment]
any comment on this?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ArithmeticExpressionSuite.scala [hunk 1]
```diff
@@ -405,6 +405,21 @@ class ArithmeticExpressionSuite extends SparkFunSuite with ExpressionEvalHelper
...
   }
 
+  test("SPARK-28322: IntegralDivide supports decimal type") {
+    withSQLConf(SQLConf.LEGACY_INTEGRALDIVIDE_RETURN_LONG.key -> "false") {
+      checkEvaluation(IntegralDivide(Literal(Decimal(1)), Literal(Decimal(2))), Decimal(0))
+      checkEvaluation(IntegralDivide(Literal(Decimal(1.4)), Literal(Decimal(0.6))), Decimal(2))
+      checkEvaluation(IntegralDivide(Literal(Decimal(1.2)), Literal(Decimal(1.1))), Decimal(1))
+      checkEvaluation(IntegralDivide(Literal(Decimal(1.2)), Literal(Decimal(0.0))), null)
+    }
+    withSQLConf(SQLConf.LEGACY_INTEGRALDIVIDE_RETURN_LONG.key -> "true") {
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ArithmeticExpressionSuite.scala]: We need to support decimals for this divide in the legacy mode, too?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala [hunk 2]
```diff
@@ -174,6 +174,20 @@ object DecimalPrecision extends TypeCoercionRule {
...
+    case expr @ IntegralDivide(
+        e1 @ DecimalType.Expression(p1, s1), e2 @ DecimalType.Expression(p2, s2)) =>
+      val widerType = widerDecimalType(p1, s1, p2, s2)
+      val promotedExpr =
+        IntegralDivide(promotePrecision(e1, widerType), promotePrecision(e2, widerType))
+      if (expr.dataType.isInstanceOf[DecimalType]) {
+        // This follows division rule
+        val intDig = p1 - s1 + s2
+        // No precision loss can happen as the result scale is 0, so only overflow can happen
+        CheckOverflow(promotedExpr, DecimalType.bounded(intDig, 0), nullOnOverflow)
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala]: Any test for the overflow case?

**Comment by maropu** [comment]
Could you uncoment the test in SQLQueryTestSuite? https://github.com/apache/spark/blob/master/sql/core/src/test/resources/sql-tests/inputs/pgSQL/numeric.sql#L897

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala [hunk 3]
```diff
@@ -394,25 +397,47 @@ case class Divide(left: Expression, right: Expression) extends DivModLike {
 // scalastyle:on line.size.limit
 case class IntegralDivide(left: Expression, right: Expression) extends DivModLike {
 
-  override def inputType: AbstractDataType = IntegralType
+  override def inputType: AbstractDataType = TypeCollection(IntegralType, DecimalType)
+
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala]: (This comment is not related to this pr though...) We need to update the description for IntegralDivide: https://github.com/apache/spark/pull/25136/files#diff-1516b10738479bbe190fb4e239258473R390

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala [hunk 4]
```diff
@@ -292,6 +292,8 @@ case class Multiply(left: Expression, right: Expression) extends BinaryArithmeti
 // Common base trait for Divide and Remainder, since these two classes are almost identical
 trait DivModLike extends BinaryArithmetic {
 
+  def decimalToDataTypeCodeGen(decimalResult: String): String = decimalResult
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala]: protected?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ArithmeticExpressionSuite.scala (hunk 1 above)
Comment (mgaido91) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ArithmeticExpressionSuite.scala]: I think it is good for consistency

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala (hunk 2 above)
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala]: it is very hard to cause an overflow, since the scale is 0, so the overflow can happen only in the `promotePrecision` phase, in those casts... maybe we can get rid of this `CheckOverflow`, although I think for safety it is good to have it. WDYT?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala]: For example `select div(-9223372036854775808, -1)` should cause overflow with `spark.sql.legacy.integralDivide.returnBigint=true`? But, in case of `spark.sql.legacy.integralDivide.returnBigint=false`, IIUC we have no chance to hit overflow. Actually, I think we should promote it without overflow.

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala]: when `spark.sql.legacy.integralDivide.returnBigint=true` anyway, the overflow should be checked by the `toLong` method, ie. the "cast" to long, which is yet to be handled in a different PR.

Let me remove this `CheckOverflow` then, thanks.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala]: > when spark.sql.legacy.integralDivide.returnBigint=true anyway, the overflow should be checked by the toLong method, ie. the "cast" to long, which is yet to be handled in a different PR.

Yea, it looks ok to me. Thanks for the check!

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala]: > Eg, if the operands are decimal(35, 0) and decimal(5, 3), they are promoted to decimal(35, 3), but the correct return type should be decimal(38,0)

Ah, I see. Thanks for the explanation! I missed that case. But, even if you know that condition, you cannot write a test case for that?

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala]: > Yea, it looks ok to me. Thanks for the check!

thank you for the comment and checking this out.

> even if you know that condition, you cannot write a test case for that?

I can, the point is that we are not testing any overflow. But I can add a test like the one mentioned above and I am doing it, thanks.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala]: Many Thanks!!

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala]: Thank you!

**Review by maropu — COMMENTED** [review]
I left some minor comments and it almost looks ok to me. cc: @cloud-fan @dongjoon-hyun @HyukjinKwon

**Comment by cloud-fan** [comment]
I'm totally fine with adding an internal config to change some behaviors to match Postgres, so that we can leverage the tests from Postgres.

However, if we want to officially change a behavior, we need to do more studies. Does SQL standard define the behavior of `div`? What's the behavior of `div` in mainstream databases? Do we have some valid use cases that need to change the behavior of `div`?

**Comment by mgaido91** [comment]
MySQL, Oracle and Postgres supports decimal as input type for `div`. I haven't been able to find this function in SQLServer.

Actually, I am not sure why you are stating that this is a behavior change (but maybe it is just wording). I mean, here we are supporting a datatype which was not supported before, so to me it seems a new use-case supported, rather than a behavior change. But we can add a config flag for this if you think it is necessary. To me, it seems a bit overkill as I see no harm is supporting an operation which is not supported otherwise and I'd see no reason to turn off that flag.

**Comment by cloud-fan** [comment]
Well, we can say this is a new feature if you think "behavior change" is too strong.

I don't like the statement that "MySQL, Oracle and Postgres support it, so Spark should support it as well". I want to see a story that drives this feature request.

In Spark, we have `/` and `div`. `/` is for fractional division and `div` is for integral division. I think this is a good story. If we make `div` support decimal, then we create confusion: what operator users should pick if they want to divide 2 decimals?

For example, "`div` should just be an alias of `/`" is also a good story to me, if that's what happened in MySQL, Oracle and Postgres.

**Comment by mgaido91** [comment]
thank for clarifying @cloud-fan, now I get your point. So let me explain a bit better:

 - `div` performs integral division, ie. it returns the integer part of a division. This is true for all the data types which are passes as operands. So it is not an alias for `/` it is a different operation with a different result;
 - currently in Spark we do not support DECIMAL operands for it (before this PR. So basically, in order to get the equivalent operation, a user needs to cast the operands to DOUBLE and then perform div on them. After this PR, these 2 casts for the operands are not needed anymore and the same operation is available also for DECIMALs.

Does this answer? Thanks.

**Comment by cloud-fan** [comment]
The story SGTM, thanks!

I need to check some details:
1. Does `div` support double/float?
2. what's the semantic? Is it equal to casting operands to long and then divide, or applying fractional division and casting the result to long?

**Comment by mgaido91** [comment]
> Does div support double/float?

no, in Postgres at least it doesn't.

> what's the semantic? 

it returns the integral part of the division. It'd go for:

> applying fractional division and casting the result to long

as far as the cast means flooring the result.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ArithmeticExpressionSuite.scala [hunk 5]
```diff
@@ -405,6 +406,28 @@ class ArithmeticExpressionSuite extends SparkFunSuite with ExpressionEvalHelper
     assert(ctx2.inlinedMutableStates.size == 1)
   }
 
+  test("SPARK-28322: IntegralDivide supports decimal type") {
+    withSQLConf(SQLConf.LEGACY_INTEGRALDIVIDE_RETURN_LONG.key -> "false") {
+      checkEvaluation(IntegralDivide(Literal(Decimal(1)), Literal(Decimal(2))), Decimal(0))
+      checkEvaluation(IntegralDivide(Literal(Decimal(1.4)), Literal(Decimal(0.6))), Decimal(2))
+      checkEvaluation(IntegralDivide(Literal(Decimal(1.2)), Literal(Decimal(1.1))), Decimal(1))
+      checkEvaluation(IntegralDivide(Literal(Decimal(1.2)), Literal(Decimal(0.0))), null)
+      checkEvaluation(DecimalPrecision.decimalAndDecimal.apply(IntegralDivide(
```
[the remark this answers is not in the document]
Comment (mgaido91) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ArithmeticExpressionSuite.scala]: why? this doesn't overflow. I don't see the reason for that

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ArithmeticExpressionSuite.scala [hunk 6]
```diff
@@ -405,6 +406,28 @@ class ArithmeticExpressionSuite extends SparkFunSuite with ExpressionEvalHelper
...
+        Literal(Decimal("99999999999999999999999999999999999")), Literal(Decimal(0.001)))),
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ArithmeticExpressionSuite.scala]: if we add more digits, can this overflow?

Comment (mgaido91) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ArithmeticExpressionSuite.scala]: No, we cannot overflow in this operation. Please see the code comment at https://github.com/apache/spark/pull/25136#discussion_r315016848

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala [hunk 7]
```diff
@@ -174,6 +174,23 @@ object DecimalPrecision extends TypeCoercionRule {
...
 
+    case expr @ IntegralDivide(
+        e1 @ DecimalType.Expression(p1, s1), e2 @ DecimalType.Expression(p2, s2)) =>
+      val widerType = widerDecimalType(p1, s1, p2, s2)
+      val promotedExpr =
+        IntegralDivide(promotePrecision(e1, widerType), promotePrecision(e2, widerType))
+      if (expr.dataType.isInstanceOf[DecimalType]) {
+        // This follows division rule
+        val intDig = p1 - s1 + s2
+        // No precision loss can happen as the result scale is 0.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala]: I don't quite get it. How can we avoid overflow for division? e.g. `div(999..., 0.000...1)`, we can't represent the result with decimal.

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala]: Because in a case like you mentioned, the overflow doesn't happen in the `CheckOverflow`, but it happens in the `PromotePrecision` on the operands. If no overflows happen on the operands, since the result has scale 0, we cannot have more than 38 digits of the integer part and no overflow can happen.

I mean, either there is an overflow in the promote precision of the operands or no overflow can happen at all.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala]: do we have a test case for operand overflow?

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala]: no, I am adding it, thanks.

**Review by cloud-fan — APPROVED** [review]

**Review by maropu — APPROVED** [review]

**Comment by maropu** [comment]
Thanks, @mgaido91! Merged to master.

**Comment by mgaido91** [comment]
thank you all guys!

**Review by HyukjinKwon — COMMENTED** [review]
Thanks for cc'ing me @maropu - I happened to check this comment later ..

LGTM too
